application/frontend/controllers/MfaController.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace frontend\controllers;
+
+use frontend\components\BaseRestController;
+use Sil\Idp\IdBroker\Client\IdBrokerClient;
+use yii\filters\AccessControl;
+use yii\helpers\ArrayHelper;
+use yii\web\BadRequestHttpException;
+
+class MfaController extends BaseRestController
+{
+
+    /**
+     * @var IdBrokerClient
+     */
+    public $idBrokerClient;
+
+    public function behaviors()
+    {
+        return ArrayHelper::merge(parent::behaviors(), [
+            'access' => [
+                'class' => AccessControl::className(),
+                'rules' => [
+                    [
+                        'allow' => true,
+                        'roles' => ['@'],
+                    ],
+                ]
+            ]
+        ]);
+    }
+
+    public function init()
+    {
+        parent::init(); // TODO: Change the autogenerated stub
+        $config = \Yii::$app->params['mfa'];
+        $this->idBrokerClient = new IdBrokerClient(
+            $config['baseUrl'],
+            $config['accessToken'],
+            [
+                IdBrokerClient::TRUSTED_IPS_CONFIG              => $config['validIpRanges']       ?? [],
+                IdBrokerClient::ASSERT_VALID_BROKER_IP_CONFIG   => $config['assertValidBrokerIp']   ?? true,
+            ]
+        );
+    }
+
+    public function actionIndex()
+    {
+        try {
+            return $this->idBrokerClient->mfaList(\Yii::$app->user->identity->employee_id);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionCreate()
+    {
+        $type = \Yii::$app->request->getBodyParam('type');
+        if ($type === null) {
+            throw new BadRequestHttpException('Type is required');
+        }
+
+        $label = \Yii::$app->request->getBodyParam('label');
+
+        try {
+            return $this->idBrokerClient->mfaCreate(\Yii::$app->user->identity->employee_id, $type, $label);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionDelete($mfaId)
+    {
+        try {
+            return $this->idBrokerClient->mfaDelete($mfaId, \Yii::$app->user->identity->employee_id);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionVerify($mfaId)
+    {
+        $value = \Yii::$app->request->getBodyParam('value');
+        if ($value === null) {
+            throw new BadRequestHttpException('Value is required');
+        }
+
+        try {
+            if ($this->idBrokerClient->mfaVerify($mfaId, \Yii::$app->user->identity->employee_id, $value)) {
+                \Yii::$app->response->statusCode = 204;
+                return;
+            }
+            throw new BadRequestHttpException('Invalid code provided');
+        } catch (\Exception $e) {
+            throw new BadRequestHttpException($e->getMessage());
```

Comment:
Rather than a `BadRequestHttpException`, should this be a server error exception of some sort (for an HTTP 500 response)?

Reply:
Also, are we sure we want to return the message of an unknown exception to the end user's browser? I know we (at least sometimes) sanitize those, logging the full error message but sending something "safe" to the end user.

Reply:
yeah that is a good idea, I'll make those changes

Reply:
How is that update?